src/app/main.rs

```diff
@@ -190,172 +198,209 @@ where
+                .and_then(|s| s.controller_identity.clone().map(|id| id));
+
+            let control_config = control_host_and_port.map(|host_and_port| {
+                control::Config::new(
+                    host_and_port,
+                    tls_server_identity,
+                    config.control_backoff_delay,
+                    config.control_connect_timeout,
+                )
             });
 
-        let (resolver, resolver_bg) = control::destination::new(
-            dns_resolver.clone(),
-            config.namespaces.clone(),
-            control_host_and_port,
-            controller_tls,
-            config.control_backoff_delay,
-            config.destination_concurrency_limit,
-        );
-
-        const MAX_IN_FLIGHT: usize = 10_000;
+            // TODO metrics
+            let stack = connect::Stack::new()
+                .push(control::client::Layer::new())
+                .push(control::resolve::Layer::new(dns_resolver.clone()))
+                .push(reconnect::layer().with_fixed_backoff(config.control_backoff_delay))
+                .push(proxy::timeout::layer(config.control_connect_timeout))
+                .push(svc::watch::layer(tls_client_config.clone()))
+                .push(svc::stack::phantom_data::layer())
+                .push(control::add_origin::Layer::new())
+                .push(buffer::layer())
+                .push(limit::layer(config.destination_concurrency_limit));
 
-        let (drain_tx, drain_rx) = drain::channel();
+            // Because the control client is buffered, we need to be able to
+            // spawn a task on an executor. An executor is only available in the
+            // context of a running task, however, we the stack is instantiated lazily.
+            future::poll_fn(move || {
+                let svc = control_config
+                    .as_ref()
+                    .and_then(|ref c| match stack.make(c) {
+                        Ok(svc) => Some(svc),
+                        Err(e) => {
+                            error!("ignoring invalid controller configuration: {:?}", e);
+                            None
+                        }
+                    });
+                Ok(svc.into())
+            })
+        };
 
-        let outbound = {
-            use super::outbound::{discovery::Resolve, orig_proto_upgrade, Endpoint, Recognize};
-            use proxy::{
-                http::{balance, metrics},
-                resolve,
+        let (resolver_bg_tx, resolver_bg_rx) = futures::sync::oneshot::channel();
+
+        // Build the outbound and inbound proxies using the controller client.
+        let main_fut = controller_fut.and_then(move |controller| {
+            let (resolver, resolver_bg) = control::destination::new(
+                controller.clone(),
+                dns_resolver.clone(),
+                config.namespaces.clone(),
+                config.destination_concurrency_limit,
+            );
+            resolver_bg_tx
```

Comment:
Is there a reason `main_fut` can't just `tokio::spawn(resolver_bg)`, rather than having to send it on the channel? This seems unnecessarily complex at a glance.

Reply:
The intent is so that resolver_bg runs on the other core.

Reply:
Ah, nevermind; this does seem like the best way to do that. I forgot we were still running the separate admin runtime.